File: tools/ModifyGpsDate.rb
#!/usr/bin/env ruby

# Script for changing the date of Gps entres in a database.
#
# Usage e.g $ ruby ModifyGpsDate.rb test.db '2011-01-28'"
#
# Peter Vasil
# Date: 2011-01-28

require 'rubygems'

require 'sqlite3'

# check if number of arguments is ok
unless ARGV.length == 2
  puts "Not the right number of arguments."
  puts "Usage: ruby ModifyGpsDate.rb test.db '2011-01-28'\n"
  exit
end

database_path = ARGV[0]
date_to_use = ARGV[1]


db = SQLite3::Database.new(database_path)

# Get all user ids and names
users = db.execute( "SELECT * FROM user" )

puts "Number of users: #{users.length}"

# date_to_use = "2011-03-19"

# Create all arrays
datetimes = Array.new(Array.new)
datetimes_new = Array.new(Array.new)
user_ids = Array.new

# Get all timestamps for all users
users.each do |user|

  datetime = Array.new
  # puts "User: #{user[0]}, name: #{user[1]}"
  db.execute("SELECT * FROM gpsdata WHERE user = ? ORDER BY datetime(time) ASC", user[0]).each do |row|
    datetime.push(row[5])
  end

  datetime_new = Array.new(datetime.length)

  datetime.length.times do |i|
    index_t = datetime[i].index('T');
    current_date = datetime[i]
    datetime_new[i] = ""
    datetime_new[i] << date_to_use
    datetime_new[i] << current_date.slice(index_t..current_date.length)
  end

  datetimes.push(datetime)
  datetimes_new.push(datetime_new)
  user_ids.push(user[0].to_s)
end

puts "Modifying dates..."

num_modified = 0

# update database entries 
db.transaction
datetimes.length.times do |i|
  datetimes[i].length.times do |j|
    db.execute("UPDATE gpsdata SET time = :newdate WHERE time = :olddate AND user = :user",
               "newdate" => datetimes_new[i][j],
               "olddate" => datetimes[i][j],
               "user" => user_ids[i])

    num_modified += 1
    
  end
end
db.commit
db.close

puts "Modified #{num_modified} entries!"

File: scripts/ModifyGpsDate.py
#!/usr/bin/python
# sqlite3 example

from optparse import OptionParser
import sqlite3
import sys
import os.path

class Database:
    def __init__(self, dbpath):
        try:
            self.conn = sqlite3.connect(dbpath)
        except sqlite3.OperationalError: # Can't locate database file
            exit(1)
        self.cursor = self.conn.cursor()
        
    def get_users(self):
        users = []
        cmd = "SELECT * FROM user"
        self.cursor.execute(cmd)
        results = self.cursor.fetchall()
        for user in results:
            users.append(user[0])

        return users
    
    def get_timestamps(self, users):
        datetimes = []
        for user in users:
            timestamps = []
            cmd = """SELECT * FROM gpsdata WHERE user = "%s" ORDER BY datetime(time) ASC """ %(user)
            self.cursor.execute(cmd)
            results = self.cursor.fetchall()
            for row in results:
                timestamps.append(row[5])
            datetimes.append(timestamps)
            
        return datetimes

    def change_dates(self, users, olddates, newdate):
        newdates = []
        for olddate in olddates:
            date_new = []
            for date in olddate:
                index_t = date.find("T", 0)
                time = date[index_t:len(date)]
                new_datetime = newdate + time
                date_new.append(new_datetime)
            newdates.append(date_new)
        
        num_processed = 0
        for i in range(len(users)):
            for j in range(len(olddates[i])):
                cmd = """UPDATE gpsdata SET time = "%s" WHERE time = "%s" AND user = %d """ %(newdates[i][j],olddates[i][j],users[i] )
                self.cursor.execute(cmd)
                num_processed += 1
        
        self.conn.commit()
        return num_processed

    def close_handle(self):
        'Closes the connection to the database'
        self.conn.commit() # Make sure all changes are saved
        self.conn.close()

def main():
    usage = "usage: %prog 'database' 'new date'"
    parser = OptionParser(usage, version="%prog 0.1")
    (options, args) = parser.parse_args()
    if len(args) != 2:
        parser.error("\nplease enter the the database and the new date")
    
    dbpath, new_date = args
    
    if not(os.path.isfile(dbpath)):
        print "database file does not exist"
        exit(1)

    db = Database(dbpath)
    
    users = db.get_users()
    print "got %d users" % len(users)
    
    datetimes = db.get_timestamps(users)
    
    print "modifying database..."
    processed_entries = db.change_dates(users,datetimes, new_date)
    
    db.close_handle()
    print "processed %d entries" % processed_entries

if __name__ == '__main__':
    sys.exit(main())
